Add dice example

pellared opened the pull request:
Towards https://github.com/open-telemetry/opentelemetry-go/issues/4531

Mostly based on https://opentelemetry.io/docs/instrumentation/python/getting-started/ (but also looked at other languages)

pellared commented:
@svrnm @cartermp PTAL

dmathieu commented on example/dice/main.go:
```diff
@@ -0,0 +1,93 @@
+	}
+	// Handle shutdown properly so nothing leaks.
+	defer func() {
+		err = errors.Join(err, otelShutdown(context.Background()))
```
This error seems to never be handled?

pellared replied:
Named return 😉

dmathieu replied:
Yes, but the `return` (on line 84) specifies its own error, it doesn't use the named variable.

pellared replied:
It will work. See:  https://go.dev/play/p/QM3oLhEKkTl

But maybe it will be better to refactor line 84 to:

```go
	err = srv.Shutdown(context.Background())
```

WDYT?

pellared replied:
https://github.com/open-telemetry/opentelemetry-go/pull/4539/commits/d87f3431b52f5f404e2318c89e25dbe46736d227

dmathieu replied:
Your latest commit makes it easier indeed, though I think it's a bit too magical (not to have things happening in-line).

pellared replied:
What do you propose instead? I do not think there is any other option than using "named returns" to alter the error value returned in defer statement.

pellared replied:
> And the return would be errors.Join(errs...)?

The returned `errors.Join(errs...)` would still miss the error returned by deferred shutdown.

pellared replied:
For reference: https://wstrm.dev/posts/errors-join-heart-defer/

MrAlias commented on example/dice/doc.go:
```diff
@@ -0,0 +1,16 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// dice is the "Roll the dice" getting started example application.
```
```suggestion
// Package main provides a "Roll the dice" getting started example application.
```

pellared replied:
Commands (`package main`) have different convention for comments. See: https://go.dev/doc/comment

pellared replied:
https://github.com/open-telemetry/opentelemetry-go/pull/4539/commits/30cd9e8f9512ca885cc66b428dc8bda5f601ace9

MrAlias commented on example/dice/otel.go:
```diff
@@ -0,0 +1,118 @@
+)
+
+// setupOTelSDK bootstraps the OpenTelemetry pipeline.
+// If it does not return an error, ake sure to call shutdown for proper cleanup.
```
```suggestion
// If it does not return an error, make sure to call shutdown for proper cleanup.
```

MrAlias commented on example/dice/main.go:
```diff
@@ -0,0 +1,104 @@
+	}
+	// Handle shutdown properly so nothing leaks.
+	defer func() {
+		err = errors.Join(err, otelShutdown(context.Background()))
```
This should check that `err` is `nil` before overwriting it here.

pellared replied:
No need. From https://pkg.go.dev/errors#Join: 

> Any nil error values are discarded.

Also for refence: https://wstrm.dev/posts/errors-join-heart-defer/

MrAlias commented on example/dice/main.go:
```diff
@@ -0,0 +1,104 @@
+	}()
+
+	// Create HTTP server with requests' base context canceled on server shutdown.
+	rCtx, cancel := context.WithCancel(context.Background())
```
Why not use `ctx` here? You have two background context in this function.

pellared replied:
Just to double check. Should I change the code to:

```go 
	srv := &http.Server{
		Addr:         ":8080",
		BaseContext:  func(_ net.Listener) context.Context { return ctx },
		ReadTimeout:  time.Second,
		WriteTimeout: 10 * time.Second,
		Handler:      newHTTPHandler(),
	}
```

and get rid of:

```go
	// Create HTTP server with requests' base context canceled on server shutdown.
	rCtx, cancel := context.WithCancel(context.Background())

 	// (...) server creation

	srv.RegisterOnShutdown(cancel)
```

?

pellared replied:
PTAL https://github.com/open-telemetry/opentelemetry-go/pull/4539/commits/bd65fec347966317e7b46825538b66d8b12f6816

pellared commented:
https://github.com/open-telemetry/opentelemetry-go/pull/4539/commits/93a69bc5261400e8b2f42edaa852e56e4eedb5ad is addressing comments from https://github.com/open-telemetry/opentelemetry.io/pull/3310#pullrequestreview-1643131028

pellared commented:
@dashpole Can you please take a quick look as I think you already reviewed this code 😉
